.doom.d/config.org
#+TITLE:Doom Emacs Config
#+AUTHOR: Adeeb HS
#+STARTUP: showeverything
#+PROPERTY: header-args :tangle config.el

* Table of contents :toc:
- [[#instructions][Instructions]]
- [[#default-configs][Default configs]]
- [[#visual][Visual]]
  - [[#general-information][General Information]]
  - [[#theme][Theme]]
  - [[#font][Font]]
  - [[#icons][Icons]]
  - [[#configure-indentation][Configure Indentation]]
- [[#org-mode-config][Org mode config]]
  - [[#default-configs-1][Default configs]]
  - [[#remap-org][Remap org]]
- [[#remap-keys][Remap Keys]]
  - [[#esc-to-fj][Esc to fj]]
  - [[#resizing-windows][Resizing windows]]
  - [[#switch-m-x-and-eval-expression-key-maps][Switch M-x and eval Expression key maps]]
  - [[#switch-evil-beginning-of-line-and-first-non-white-space-character-keys][Switch evil beginning of line and first non white space character keys]]
  - [[#define-back-tab][Define back tab]]
- [[#devepolment][Devepolment]]
  - [[#languages][Languages]]
  - [[#competitive-coding][Competitive coding]]
- [[#company-mode][Company mode]]
- [[#browsers][Browsers]]
  - [[#eww-browser][Eww browser]]
- [[#dashboard][Dashboard]]
- [[#hide-files-or-directories][Hide files or directories]]
- [[#flyspell][FlySpell]]

* Instructions
1. Place your private configuration here! Remember, you do not need to run 'doom sync' after modifying this file!

2. Here are some additional functions/macros that could help you configure Doom:
   + ~load!~ for loading external *.el files relative to this one
   + ~use-package!~ for configuring packages
   + ~after!~ for running code after a package has loaded
   + ~add-load-path!~ for adding directories to the ~load-path~, relative to this file. Emacs searches the ~load-path~ when you load packages with ~require~ or ~use-package~.
   + ~map!~ for binding new keys

3. To get information about any of these functions/macros, move the cursor over the highlighted symbol at press 'K'. This will open documentation for it, including demos of how they are used.

4. You can also try 'gd' to jump to their definition and see how they are implemented.

* Default configs
#+begin_src elisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
(setq user-full-name "adeeb"
      user-mail-address "adeeb.hs1@gmail.com"
      display-line-numbers-type 'relative
      ;;confirm-kill-emacs nil
)
#+end_src

* Visual
** General Information
+ Doom exposes five (optional) variables for controlling fonts in Doom. Here are the three important ones:
  - ~doom-font~
  - ~doom-variable-pitch-font~
  - ~doom-big-font~ -- used for ~doom-big-font-mode~; use this for presentations or streaming.

+ They all accept either a font-spec, font string ("Input Mono-12"), or xlfd font string. You generally only need these two:

+ There are two ways to load a theme. Both assume the theme is installed and available. You can either set ~doom-theme~ or manually load a theme with the ~load-theme~ function. This is the default:

** Theme
#+begin_src elisp
(setq doom-theme 'doom-dracula)
#+end_src

** Font
#+begin_src elisp
(setq doom-font (font-spec :family "monospace" :size 25 :weight 'semi-light)
       doom-variable-pitch-font (font-spec :family "sans" :size 23))

(use-package fira-code-mode
  :config (fira-code-mode-set-font)
  :custom (fira-code-mode-disabled-ligatures '("[]" "#{" "#(" "#_" "#_(" "x" "***" "<>")) ;; List of ligatures to turn off
  :hook prog-mode org-mode) ;; Enables fira-code-mode automatically for programming  and org major modes
#+end_src

** Icons
#+begin_src elisp
(use-package all-the-icons
  :if (display-graphic-p))
(use-package all-the-icons-dired
  :hook (dired-mode . all-the-icons-dired-mode)
  )             ;; Icons for dired
(setq doom-themes-treemacs-theme "doom-colors")
(setq +ligatures-extras-in-modes '(haskell-mode org-mode))
(set-scroll-bar-mode 'right)
#+end_src

** Configure Indentation
#+begin_src elisp
(use-package highlight-indent-guides
  :ensure t
  :commands highlight-indent-guides-mode
  :hook (prog-mode . highlight-indent-guides-mode)
  :config
  (setq highlight-indent-guides-method 'character
        ;;highlight-indent-guides-character ?\❯
   )
 )
#+end_src

* Org mode config
** Default configs
#+begin_src elisp
(setq org-directory "~/org/")
(map! :leader
      :desc "Org babel tangle" "m B" #'org-babel-tangle)
(after! org
  (setq
        org-pretty-entities t
        org-startup-folded t
        org-ellipsis " ▼ "
        org-superstar-headline-bullets-list '("◉" "●" "○" "◆" "●" "○" "◆")
        ;;org-superstar-item-bullet-alist '((?+ . ?➤) (?- . ?✦)) ; changes +/- symbols in item lists
        org-log-done 'time
        org-hide-emphasis-markers t
        )
  )

;; Shorten some text
;; (defun my/org-mode/load-prettify-symbols ()
;;   (interactive)
;;   (setq prettify-symbols-alist
;;     '(("#+begin_src" . ?)
;;       ("#+BEGIN_SRC" . ?)
;;       ("#+end_src" . ?)
;;       ("#+END_SRC" . ?)
;;       ("#+begin_example" . ?)
;;       ("#+BEGIN_EXAMPLE" . ?)
;;       ("#+end_example" . ?)
;;       ("#+END_EXAMPLE" . ?)
;;       ("#+header:" . ?)
;;       ("#+HEADER:" . ?)
;;       ("#+results:" . ?)
;;       ("#+RESULTS:" . ?)
;;       ("#+call:" . ?)
;;       ("#+CALL:" . ?)
;;       (":PROPERTIES:" . ?)
;;       (":properties:" . ?)
;;       ))
;;   (prettify-symbols-mode 1))
;; (add-hook 'org-mode-hook 'my/org-mode/load-prettify-symbols)
#+end_src

** Remap org

* Remap Keys
** Esc to fj
#+begin_src elisp
(setq key-chord-two-keys-delay 0.15)
(key-chord-define evil-insert-state-map "fj" 'evil-normal-state)
(key-chord-mode 1)
#+end_src

** Resizing windows
Use ~SPC w c~ and then the following keys

| Description            | Key Binding |
|------------------------+-------------|
| Increase Window Width  | =           |
| Decrease Window Width  | -           |
| Increase Window Height | +           |
| Decrease Window Height | _           |

#+begin_src elisp
(defhydra hydra-window-size (:timeout 5)
"Resize window"
("=" evil-window-increase-width "increaseW")
("-" evil-window-decrease-width "decreaseW")
("+" evil-window-increase-height "increaseH")
("_" evil-window-decrease-height "decreaseH")
("f" nil "finished" :exit t))

(map! :leader
    (:prefix ("w". "window")
    :desc "Resize current window" "c" #'hydra-window-size/body))
#+end_src

** Switch M-x and eval Expression key maps

#+begin_src elisp
(map! :leader
      :desc "M-x" ";" #'execute-extended-command)

(map! :leader
      :desc "Eval Expression" ":" #'eval-expression)
#+end_src

** Switch evil beginning of line and first non white space character keys
#+begin_src elisp
(define-key evil-normal-state-map (kbd "0") #'evil-first-non-blank)
(define-key evil-normal-state-map (kbd "^") #'evil-beginning-of-line)
#+end_src

** Define back tab

#+begin_src elisp
(global-set-key (kbd "<backtab>") 'un-indent-by-removing-4-spaces)
(defun un-indent-by-removing-4-spaces ()
  "remove 4 spaces from beginning of of line"
  (interactive)
  (save-excursion
    (save-match-data
      (beginning-of-line)
      ;; get rid of tabs at beginning of line
      (when (looking-at "^\\s-+")
        (untabify (match-beginning 0) (match-end 0)))
      (when (looking-at "^    ")
        (replace-match "")))))
#+end_src

* Devepolment
** Languages
*** Language Server
**** Lsp Mode
#+begin_src elisp
(defun efs/lsp-mode-setup ()
  (setq lsp-headerline-breadcrumb-segments '(path-up-to-project file symbols))
  (lsp-headerline-breadcrumb-mode))

(use-package lsp-mode
  :after lsp
  :commands (lsp lsp-deferred)
  :hook (lsp-mode . efs/lsp-mode-setup)
  :config
  (lsp-enable-which-key-integration t))
#+end_src

**** Lsp Ui
#+begin_src elisp
(use-package lsp-ui
  :after lsp
  :hook (lsp-mode . lsp-ui-mode)
  :custom
  (lsp-ui-doc-position 'bottom)
  )

#+end_src

**** Lsp Treemacs
#+begin_src elisp
(use-package lsp-treemacs
  :after lsp)
#+end_src

*** C/C++
#+begin_src elisp
;;(setq lsp-clients-clangd-args '("--header-insertion=never"))

#+end_src

*** Python
**** Set-up LSP for python
#+begin_src elisp
(use-package lsp-pyright
  :after lsp
  :ensure t
  :init
  (setq lsp-pyright-multi-root nil)
  :hook (python-mode . (lambda ()
                          (require 'lsp-pyright)
                          (lsp))))  ; or lsp-deferred
#+end_src

*** Haskell
**** Setting up extra ligatures just for haskell
#+begin_src elisp
(plist-put! +ligatures-extra-symbols
            :sum        "Σ"
            :product    "Ⲡ"
            )

(after! haskell-mode
  (set-ligatures! 'haskell-mode
    :lambda        "\\"
    :composition   "."
    :null          "()"
    :int           "Int"
    :float         "Double"
    :str           "String"
    :bool          "Bool"
    :in            "`elem`"
    :not-in        "`notElem`"
    :union         "`union`"
    :intersect     "`intersect`"
    :or            "||"
    :and           "&&"
    :for           "forall"
    :sum           "sum"
    :product       "product"
    )
   )

#+end_src

** Competitive coding
*** Add c++ template automatically while file creation
#+begin_src emacs-lisp
(use-package autoinsert
  :config
  (setq auto-insert-query nil)             ; disable the default auto-inserts
  (auto-insert-mode 1)                     ; enable auto-insert-mode globally
  (add-hook 'find-file-hook 'auto-insert)  ; insert templates when we create new files
  (setq auto-insert-alist nil)             ; remove this line to restore defaults
  ;; (add-to-list 'auto-insert-alist          ; add "competitive coding" templates to auto insert
  ;;              '("^/home/adeeb/code/.+\\.cpp\\'" . "/home/adeeb/code/template.cpp"))
  (add-to-list 'auto-insert-alist          ; the same with ~ expansion
               (cons (concat "^" (expand-file-name "~/code/") ".+\\.cpp\\'")
                     (expand-file-name "~/code/template.cpp")))
 )

#+end_src

* Company mode
#+begin_src elisp
(use-package company
  :after lsp-mode
  :hook (lsp-mode . company-mode)
  ;; :bind (:map company-active-map
  ;;        ("<tab>" . company-complete-selection))
  ;;       (:map lsp-mode-map
  ;;        ("<tab>" . company-indent-or-complete-common))
  :custom
  (company-minimum-prefix-length 1)
  (company-idle-delay 0.0))

(use-package company-box
  :hook (company-mode . company-box-mode))

#+end_src

* Browsers
** Eww browser
#+begin_src elisp
(setq
 ;; browse-url-browser-function 'eww-browse-url                    ; Use eww as the default browser
 shr-use-fonts  nil                                             ; No special fonts
 shr-use-colors nil                                             ; No colours
 shr-indentation 2                                              ; Left-side margin
 shr-width 70                                                   ; Fold text to 70 columns
 shr-image-animate nil                                          ; Amination switched off
 shr-inhibit-images t                                           ; Images are switched off
 )
(defun my/eww-toggle-images ()
  "Toggle whether images are loaded and reload the current page from cache."
  (interactive)
  (setq-local shr-inhibit-images (not shr-inhibit-images))
  (eww-reload t)
  (message "Images are now %s"
           (if shr-inhibit-images "off" "on")))
#+end_src

* Dashboard
Basic set up for dashboard
#+begin_src elisp
(use-package dashboard
  :init
  (setq dashboard-set-heading-icons t)
  (setq dashboard-set-file-icons t)
  (setq dashboard-banner-logo-title nil)
  (setq dashboard-startup-banner 'logo) ;; use standard emacs logo as banner
  ;;(setq dashboard-startup-banner "~/.doom.d/doom-emacs-dash.png")  ;; use custom image as banner
  (setq dashboard-center-content nil) ;; set to 't' for centered content
  (setq dashboard-items '((recents . 5)
                          (projects . 5)))

  :custom
        (dashboard-set-init-info t)
        (dashboard-set-navigator t)
        (dashboard-projects-backend 'projectile)

:config
  (dashboard-setup-startup-hook)
  (dashboard-modify-heading-icons '((bookmarks . "book"))))

#+end_src

* Hide files or directories
#+begin_src elisp
(add-to-list 'recentf-exclude "/.emacs.d/.local/etc/workspaces/autosave") ;;hide recent files from recentf
(add-to-list 'projectile-ignored-projects "~/.emacs.d")                 ;;hide emacs.d dir from projectile projects
#+end_src

* FlySpell
Spell checker for text mode buffers
#+begin_src elisp
(use-package flyspell
  :ensure nil
  :diminish
  :if (executable-find "aspell")
  :hook (((text-mode outline-mode latex-mode org-mode markdown-mode) . flyspell-mode))
  :custom
  (flyspell-issue-message-flag nil)
  (ispell-program-name "aspell")
  (ispell-extra-args
   '("--sug-mode=ultra" "--lang=en_US" "--camel-case"))
  )

#+end_src


emacs/.doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
(setq user-full-name "adeeb"
      user-mail-address "adeeb.hs1@gmail.com"
      display-line-numbers-type 'relative
      ;;confirm-kill-emacs nil
)

(setq doom-theme 'doom-dracula)

(setq doom-font (font-spec :family "monospace" :size 25 :weight 'semi-light)
       doom-variable-pitch-font (font-spec :family "sans" :size 23))

(use-package fira-code-mode
  :config (fira-code-mode-set-font)
  :custom (fira-code-mode-disabled-ligatures '("[]" "#{" "#(" "#_" "#_(" "x" "***" "<>")) ;; List of ligatures to turn off
  :hook prog-mode org-mode) ;; Enables fira-code-mode automatically for programming  and org major modes

(use-package all-the-icons
  :if (display-graphic-p))
(use-package all-the-icons-dired
  :hook (dired-mode . all-the-icons-dired-mode)
  )             ;; Icons for dired
(setq doom-themes-treemacs-theme "doom-colors")
(setq +ligatures-extras-in-modes '(haskell-mode org-mode))
(set-scroll-bar-mode 'right)

(use-package highlight-indent-guides
  :ensure t
  :commands highlight-indent-guides-mode
  :hook (prog-mode . highlight-indent-guides-mode)
  :config
  (setq highlight-indent-guides-method 'character
        ;;highlight-indent-guides-character ?\❯
   )
 )

(setq org-directory "~/org/")
(map! :leader
      :desc "Org babel tangle" "m B" #'org-babel-tangle)
(after! org
  (setq
        org-pretty-entities t
        org-startup-folded t
        org-ellipsis " ▼ "
        org-superstar-headline-bullets-list '("◉" "●" "○" "◆" "●" "○" "◆")
        ;;org-superstar-item-bullet-alist '((?+ . ?➤) (?- . ?✦)) ; changes +/- symbols in item lists
        org-log-done 'time
        org-hide-emphasis-markers t
        )
  )

;; Shorten some text
;; (defun my/org-mode/load-prettify-symbols ()
;;   (interactive)
;;   (setq prettify-symbols-alist
;;     '(("#+begin_src" . ?)
;;       ("#+BEGIN_SRC" . ?)
;;       ("#+end_src" . ?)
;;       ("#+END_SRC" . ?)
;;       ("#+begin_example" . ?)
;;       ("#+BEGIN_EXAMPLE" . ?)
;;       ("#+end_example" . ?)
;;       ("#+END_EXAMPLE" . ?)
;;       ("#+header:" . ?)
;;       ("#+HEADER:" . ?)
;;       ("#+results:" . ?)
;;       ("#+RESULTS:" . ?)
;;       ("#+call:" . ?)
;;       ("#+CALL:" . ?)
;;       (":PROPERTIES:" . ?)
;;       (":properties:" . ?)
;;       ))
;;   (prettify-symbols-mode 1))
;; (add-hook 'org-mode-hook 'my/org-mode/load-prettify-symbols)

(setq key-chord-two-keys-delay 0.15)
(key-chord-define evil-insert-state-map "fj" 'evil-normal-state)
(key-chord-mode 1)

(defhydra hydra-window-size (:timeout 5)
"Resize window"
("=" evil-window-increase-width "increaseW")
("-" evil-window-decrease-width "decreaseW")
("+" evil-window-increase-height "increaseH")
("_" evil-window-decrease-height "decreaseH")
("f" nil "finished" :exit t))

(map! :leader
    (:prefix ("w". "window")
    :desc "Resize current window" "c" #'hydra-window-size/body))

(map! :leader
      :desc "M-x" ";" #'execute-extended-command)

(map! :leader
      :desc "Eval Expression" ":" #'eval-expression)

(define-key evil-normal-state-map (kbd "0") #'evil-first-non-blank)
(define-key evil-normal-state-map (kbd "^") #'evil-beginning-of-line)

(global-set-key (kbd "<backtab>") 'un-indent-by-removing-4-spaces)
(defun un-indent-by-removing-4-spaces ()
  "remove 4 spaces from beginning of of line"
  (interactive)
  (save-excursion
    (save-match-data
      (beginning-of-line)
      ;; get rid of tabs at beginning of line
      (when (looking-at "^\\s-+")
        (untabify (match-beginning 0) (match-end 0)))
      (when (looking-at "^    ")
        (replace-match "")))))

(defun efs/lsp-mode-setup ()
  (setq lsp-headerline-breadcrumb-segments '(path-up-to-project file symbols))
  (lsp-headerline-breadcrumb-mode))

(use-package lsp-mode
  :after lsp
  :commands (lsp lsp-deferred)
  :hook (lsp-mode . efs/lsp-mode-setup)
  :config
  (lsp-enable-which-key-integration t))

(use-package lsp-ui
  :after lsp
  :hook (lsp-mode . lsp-ui-mode)
  :custom
  (lsp-ui-doc-position 'bottom)
  )

(use-package lsp-treemacs
  :after lsp)

;;(setq lsp-clients-clangd-args '("--header-insertion=never"))

(use-package lsp-pyright
  :after lsp
  :ensure t
  :init
  (setq lsp-pyright-multi-root nil)
  :hook (python-mode . (lambda ()
                          (require 'lsp-pyright)
                          (lsp))))  ; or lsp-deferred

(plist-put! +ligatures-extra-symbols
            :sum        "Σ"
            :product    "Ⲡ"
            )

(after! haskell-mode
  (set-ligatures! 'haskell-mode
    :lambda        "\\"
    :composition   "."
    :null          "()"
    :int           "Int"
    :float         "Double"
    :str           "String"
    :bool          "Bool"
    :in            "`elem`"
    :not-in        "`notElem`"
    :union         "`union`"
    :intersect     "`intersect`"
    :or            "||"
    :and           "&&"
    :for           "forall"
    :sum           "sum"
    :product       "product"
    )
   )

(use-package autoinsert
  :config
  (setq auto-insert-query nil)             ; disable the default auto-inserts
  (auto-insert-mode 1)                     ; enable auto-insert-mode globally
  (add-hook 'find-file-hook 'auto-insert)  ; insert templates when we create new files
  (setq auto-insert-alist nil)             ; remove this line to restore defaults
  ;; (add-to-list 'auto-insert-alist          ; add "competitive coding" templates to auto insert
  ;;              '("^/home/adeeb/code/.+\\.cpp\\'" . "/home/adeeb/code/template.cpp"))
  (add-to-list 'auto-insert-alist          ; the same with ~ expansion
               (cons (concat "^" (expand-file-name "~/code/") ".+\\.cpp\\'")
                     (expand-file-name "~/code/template.cpp")))
 )

(use-package company
  :after lsp-mode
  :hook (lsp-mode . company-mode)
  ;; :bind (:map company-active-map
  ;;        ("<tab>" . company-complete-selection))
  ;;       (:map lsp-mode-map
  ;;        ("<tab>" . company-indent-or-complete-common))
  :custom
  (company-minimum-prefix-length 1)
  (company-idle-delay 0.0))

(use-package company-box
  :hook (company-mode . company-box-mode))

(setq
 ;; browse-url-browser-function 'eww-browse-url                    ; Use eww as the default browser
 shr-use-fonts  nil                                             ; No special fonts
 shr-use-colors nil                                             ; No colours
 shr-indentation 2                                              ; Left-side margin
 shr-width 70                                                   ; Fold text to 70 columns
 shr-image-animate nil                                          ; Amination switched off
 shr-inhibit-images t                                           ; Images are switched off
 )
(defun my/eww-toggle-images ()
  "Toggle whether images are loaded and reload the current page from cache."
  (interactive)
  (setq-local shr-inhibit-images (not shr-inhibit-images))
  (eww-reload t)
  (message "Images are now %s"
           (if shr-inhibit-images "off" "on")))

(use-package dashboard
  :init
  (setq dashboard-set-heading-icons t)
  (setq dashboard-set-file-icons t)
  (setq dashboard-banner-logo-title nil)
  (setq dashboard-startup-banner 'logo) ;; use standard emacs logo as banner
  ;;(setq dashboard-startup-banner "~/.doom.d/doom-emacs-dash.png")  ;; use custom image as banner
  (setq dashboard-center-content nil) ;; set to 't' for centered content
  (setq dashboard-items '((recents . 5)
                          (projects . 5)))

  :custom
        (dashboard-set-init-info t)
        (dashboard-set-navigator t)
        (dashboard-projects-backend 'projectile)

:config
  (dashboard-setup-startup-hook)
  (dashboard-modify-heading-icons '((bookmarks . "book"))))

(add-to-list 'recentf-exclude "/.emacs.d/.local/etc/workspaces/autosave") ;;hide recent files from recentf
(add-to-list 'projectile-ignored-projects "~/.emacs.d")                 ;;hide emacs.d dir from projectile projects

(use-package flyspell
  :ensure nil
  :diminish
  :if (executable-find "aspell")
  :hook (((text-mode outline-mode latex-mode org-mode markdown-mode) . flyspell-mode))
  :custom
  (flyspell-issue-message-flag nil)
  (ispell-program-name "aspell")
  (ispell-extra-args
   '("--sug-mode=ultra" "--lang=en_US" "--camel-case"))
  )
